com.unity.render-pipelines.high-definition/Runtime/Lighting/VolumetricLighting/DensityVolumeManager.cs

```diff
@@ -1,115 +1,103 @@
 
+        static DensityVolumeManager m_Manager;
         public static DensityVolumeManager manager
         {
             get
             {
-                if (_instance == null)
+                if (m_Manager == null)
+                    m_Manager = new DensityVolumeManager();
+                return m_Manager;
+            }
+        }
+
+        Texture3DAtlas m_VolumeAtlas = null;
+        public Texture3DAtlas volumeAtlas
+        {
+            get
+            {
+                if (m_VolumeAtlas == null)
                 {
-                    _instance = new DensityVolumeManager();
+                    var settings = HDRenderPipeline.currentAsset.currentPlatformRenderPipelineSettings.lightLoopSettings;
+                    m_VolumeAtlas = new Texture3DAtlas(densityVolumeAtlasFormat, (int)settings.maxDensityVolumeSize, (int)settings.maxDensityVolumesOnScreen);
+
+                    // When HDRP is initialized and this atlas created, some density volume may have been initialized before so we add them here.
+                    foreach (var volume in m_Volumes)
+                    {
+                        if (volume.parameters.volumeMask != null)
+                            AddTextureIntoAtlas(volume.parameters.volumeMask);
+                    }
                 }
-                return _instance;
+
+                return m_VolumeAtlas;
             }
         }
 
-        public Texture3DAtlas volumeAtlas = null;
-        private bool atlasNeedsRefresh = false;
-
-        //TODO: hardcoded size....:-(
-        public static int volumeTextureSize = 32;
+        List<DensityVolume> m_Volumes = null;
 
-        private DensityVolumeManager()
+        DensityVolumeManager()
         {
-            volumes = new List<DensityVolume>();
-
-            volumeAtlas = new Texture3DAtlas(TextureFormat.Alpha8, volumeTextureSize);
-
-            volumeAtlas.OnAtlasUpdated += AtlasUpdated;
+            m_Volumes = new List<DensityVolume>();
         }
 
-        private List<DensityVolume> volumes = null;
-
         public void RegisterVolume(DensityVolume volume)
         {
-            volumes.Add(volume);
-
-            volume.OnTextureUpdated += TriggerVolumeAtlasRefresh;
+            m_Volumes.Add(volume);
 
             if (volume.parameters.volumeMask != null)
             {
-                volumeAtlas.AddTexture(volume.parameters.volumeMask);
+                if (volumeAtlas.IsTextureValid(volume.parameters.volumeMask))
+                {
+                    AddTextureIntoAtlas(volume.parameters.volumeMask);
+                }
             }
         }
 
-        public void DeRegisterVolume(DensityVolume volume)
+        internal void AddTextureIntoAtlas(Texture volumeTexture)
         {
-            if (volumes.Contains(volume))
-            {
-                volumes.Remove(volume);
-            }
+            if (!volumeAtlas.AddTexture(volumeTexture))
```

Comment:
I might have missed it but does this mean the atlas is updated every frame now?  How do we detect if something has changed if not?

Reply:
This only happens when the density volume texture is registered so no. Otherwise the it's the atlas that internally checks if the textures needs to be updated with a hash:
https://github.com/Unity-Technologies/Graphics/pull/2751/files#diff-9452a7b53ae106d6e8115c2ce8966e7dec933f97ce8d7241a18130cc5d9d4af7R373